llama.cpp

```diff
@@ -5865,43 +5274,343 @@ static struct ggml_cgraph * llama_build_graph(
 
     struct ggml_cgraph * result = NULL;
 
+    // check if we should build the worst-case graph (for memory measurement)
+    const bool worst_case = ggml_allocr_is_measure(lctx.alloc);
+
     switch (model.arch) {
         case LLM_ARCH_LLAMA:
             {
-                result = llm_build_llama(lctx, batch);
+                result = llm_build_llama(lctx, batch, worst_case);
             } break;
         case LLM_ARCH_BAICHUAN:
             {
-                result = llm_build_baichaun(lctx, batch);
+                result = llm_build_baichaun(lctx, batch, worst_case);
             } break;
         case LLM_ARCH_FALCON:
             {
-                result = llm_build_falcon(lctx, batch);
+                result = llm_build_falcon(lctx, batch, worst_case);
             } break;
         case LLM_ARCH_STARCODER:
             {
-                result = llm_build_starcoder(lctx, batch);
+                result = llm_build_starcoder(lctx, batch, worst_case);
             } break;
         case LLM_ARCH_PERSIMMON:
             {
-                result = llm_build_persimmon(lctx, batch);
+                result = llm_build_persimmon(lctx, batch, worst_case);
             } break;
         case LLM_ARCH_REFACT:
             {
-                result = llm_build_refact(lctx, batch);
+                result = llm_build_refact(lctx, batch, worst_case);
             } break;
         case LLM_ARCH_BLOOM:
             {
-                result = llm_build_bloom(lctx, batch);
+                result = llm_build_bloom(lctx, batch, worst_case);
             } break;
         case LLM_ARCH_MPT:
             {
-                result = llm_build_mpt(lctx, batch);
+                result = llm_build_mpt(lctx, batch, worst_case);
             } break;
         default:
             GGML_ASSERT(false);
     }
 
+    // allocate memory and set the values for the input tensors of the graph
+
+    // inp_tokens
+    if (batch.token) {
+        struct ggml_tensor * cur = ggml_graph_get_tensor(result, "inp_tokens");
+        GGML_ASSERT(cur != nullptr);
+
+        ggml_allocr_alloc(lctx.alloc, cur);
+
+        if (!ggml_allocr_is_measure(lctx.alloc)) {
+            const int64_t n_tokens = cur->ne[0];
+
+            memcpy(cur->data, batch.token, n_tokens*ggml_element_size(cur));
+        }
+    }
+
+    // inp_embd
+    if (batch.embd) {
+        struct ggml_tensor * cur = ggml_graph_get_tensor(result, "inp_embd");
+        GGML_ASSERT(cur != nullptr);
+
+        ggml_allocr_alloc(lctx.alloc, cur);
+
+        if (!ggml_allocr_is_measure(lctx.alloc)) {
+            const int64_t n_embd   = cur->ne[0];
+            const int64_t n_tokens = cur->ne[1];
+
+            memcpy(cur->data, batch.embd, n_tokens*n_embd*ggml_element_size(cur));
+        }
+    }
+
+    // inp_pos
+    do {
+        struct ggml_tensor * cur = ggml_graph_get_tensor(result, "inp_pos");
+        if (cur == nullptr) {
+            break;
+        }
```

Comment:
A more idiomatic way to write this in C++ is:
```cpp
if (struct ggml_tensor * cur = ggml_graph_get_tensor(result, "inp_pos")) {
    // ...
}
```
Or, more verbosely:
```cpp
if (struct ggml_tensor * cur = ggml_graph_get_tensor(result, "inp_pos"); cur != nullptr) {
    // ...
}
```

Reply:
Personally, I find the declaration in a separate line easier to read. The second requires C++17, but we target C++11.

Reply:
My point is that the `do { ... break ... } while (0)` can be eliminated this way. I think it's a pattern that should be used sparingly. It may make sense as an alternative to `goto` in a function that needs to jump to a cleanup path in several places, but it looks strange to have a loop with the sole purpose of scoping a single variable.

Another option is to replace the loop with a simple `{ ... }` block, and to use a simple `if` statement inside of it, without the loop.

Reply:
I don't disagree with this, I don't like this pattern either. Personally I would prefer to move the code to a function instead if the possibility of early return is important. In this particular case, it seems to be done just to save one indentation level. Ultimately though, this is just a matter of style, and in that regard I am more concerned about consistency more than anything else (and to be fair, this pattern isn't used anywhere else in the code).